Function pointer primitive cast not working

When running the following Rust code:

```rust
fn on_event2(event: &EventA<'_>) {

}

struct EventA<'a>(&'a ());

fn main() {
    const ptr: &'static fn(&EventA<'_>) = &on_event2 as _;
}
```
https://play.rust-lang.org/?version=stable&mode=debug&edition=2024

The cast using `as _` should succeed, or the compiler should give a clearer reason for rejection if it is truly invalid. Since the types are compatible under HRLB, i guess this coercion should ideally be allowed.

However, the compiler currently emits the following error:
```
error[E0605]: non-primitive cast: `&for<'a, 'b> fn(&'a EventA<'b>) {on_event2}` as `&for<'a, 'b> fn(&'a EventA<'b>)`
 --> src/main.rs:8:43
  |
8 |     const ptr: &'static fn(&EventA<'_>) = &on_event2 as _;
  |                                           ^^^^^^^^^^^^^^^ an `as` expression can only be used to convert between primitive types or to coerce to a specific trait object

```
This happens on stable, beta, and nightly Rust.


**Workaround:**

```RUST
fn on_event2(event: &EventA<'_>) {

}

struct EventA<'a>(&'a ());

const fn force<F: Fn(&EventA<'_>)>(x: F) -> F {
    x
}

fn main() {
    const ptr: &'static fn(&EventA<'_>) = &force(|event| { on_event2(event) });
}
```

Thanks to @_madfrog on Discord for the help

--- a/compiler/rustc_hir_typeck/src/cast.rs
+++ b/compiler/rustc_hir_typeck/src/cast.rs
@@ -408,6 +408,16 @@ impl<'a, 'tcx> CastCheck<'tcx> {
                     self.expr_ty,
                     fcx.ty_to_string(self.cast_ty)
                 );
+
+                if let Ok(snippet) = fcx.tcx.sess.source_map().span_to_snippet(self.expr_span)
+                    && matches!(self.expr.kind, ExprKind::AddrOf(..))
+                {
+                    err.note(format!(
+                        "casting reference expression `{}` because `&` binds tighter than `as`",
+                        snippet
+                    ));
+                }
+
                 let mut sugg = None;
                 let mut sugg_mutref = false;
                 if let ty::Ref(reg, cast_ty, mutbl) = *self.cast_ty.kind() {

--- a/tests/ui/cast/func-pointer-issue-140491.rs
+++ b/tests/ui/cast/func-pointer-issue-140491.rs
@@ -1,0 +1,7 @@
+fn my_fn(event: &Event<'_>) {}
+
+struct Event<'a>(&'a ());
+
+fn main() {
+    const ptr: &fn(&Event<'_>) = &my_fn as _; //~ ERROR non-primitive cast: `&for<'a, 'b> fn(&'a Event<'b>) {my_fn}` as `&for<'a, 'b> fn(&'a Event<'b>)` [E0605]
+}

--- a/tests/ui/cast/func-pointer-issue-140491.stderr
+++ b/tests/ui/cast/func-pointer-issue-140491.stderr
@@ -1,0 +1,9 @@
+error[E0605]: non-primitive cast: `&for<'a, 'b> fn(&'a Event<'b>) {my_fn}` as `&for<'a, 'b> fn(&'a Event<'b>)`
+  --> $DIR/func-pointer-issue-140491.rs:6:34
+   |
+LL | ..._>) = &my_fn as _;
+   |          ^^^^^^^^^^^ an `as` expression can only be used to convert between primitive types or to coerce to a specific trait object
+
+error: aborting due to 1 previous error
+
+For more information about this error, try `rustc --explain E0605`.

--- a/tests/ui/cast/func-pointer-issue-140491.stderr
+++ b/tests/ui/cast/func-pointer-issue-140491.stderr
@@ -3,6 +3,8 @@ error[E0605]: non-primitive cast: `&for<'a, 'b> fn(&'a Event<'b>) {my_fn}` as `&
    |
 LL | ..._>) = &my_fn as _;
    |          ^^^^^^^^^^^ an `as` expression can only be used to convert between primitive types or to coerce to a specific trait object
+   |
+   = note: casting reference expression `&my_fn` because `&` binds tighter than `as`
 
 error: aborting due to 1 previous error
 

--- a/tests/ui/coercion/issue-73886.stderr
+++ b/tests/ui/coercion/issue-73886.stderr
@@ -3,6 +3,8 @@ error[E0605]: non-primitive cast: `&&[i32; 1]` as `&[_]`
    |
 LL |     let _ = &&[0] as &[_];
    |             ^^^^^^^^^^^^^ an `as` expression can only be used to convert between primitive types or to coerce to a specific trait object
+   |
+   = note: casting reference expression `&&[0]` because `&` binds tighter than `as`
 
 error[E0605]: non-primitive cast: `u32` as `Option<_>`
   --> $DIR/issue-73886.rs:4:13

